book/src/tutorial/qt-gui/qquickpainteditem.md

```diff
@@ -0,0 +1,211 @@
+    This is only necessary because we want to use CXX-Qt within the file.
+    Doing this second step is not necessary for a normal Rust module.
+```rust
+use cxx_qt_build::{CxxQtBuilder, QmlModule};
+
+fn main() {
+    CxxQtBuilder::new()
+        // - Qt Core is always linked
+        // - Qt Qml is linked by enabling the qt_qml Cargo feature.
+        // - Qt Qml requires linking Qt Network on macOS
+        .qt_module("Network")
+        .qt_module("Quick")
+        .qml_module(QmlModule {
+            uri: "com.kdab.cxx_qt.demo",
+            rust_files: &["src/image_painter.rs"], // <- Add all files that
+            qml_files: &["qml/main.qml"],          //    include a #[cxx_qt::bridge].
+            ..Default::default()
+        })
+        .build();
+}
+```
+> It is a common mistake to forget to include CXX-Qt files in the `build.rs` script.
+> Make sure to try removing the file from the list so you know what happens in this case.
+>
+> ❓ What error do you get?
+
+Now we can start adding a new `QQuickPaintedItem` subclass.
+
+CXX-Qt allows us to create new `QObject` (and therefore `QQuickPaintedItem`) subclasses.
+These subclasses are defined in C++ and their data layout cannot be changed from Rust directly.
+Instead, every subclass generated by CXX-Qt contains an instance of a Rust struct.
+It is this Rust struct that therefore defines the data layout used within CXX-Qt objects.
+
+We can create a new QObject class for any Rust struct we like.
+
+✅ Create a new empty Rust struct for our class\
+
+The struct should implement the `Default` trait.
+Ideally name the struct `ImagePainterRust`.
+
+<details>
+<summary>Example definition</summary>
+
+```rust
+#[derive(Default)]
+struct ImagePainterRust {}
+```
+</details>
+
+Now we can create our `QQuickPaintedItem` subclass by using a `#[cxx_qt::bridge]` module.
+
+This bridge module is similar to a header file in C++.
+It only includes declarations, not definitions.
+Using the bridge we can declare which items exist in C++ and expose them to Rust.
+And we can also declare Rust items to be exposed to C++.
+
+The syntax is based on and compatible with CXX.
+You can read up on CXX in the [CXX book][CXX].
+
+To create our `QQuickPaintedItem` subclass, we first need to include the appropriate C++ headers,
+then declare the types we want to use from Rust.
+For our first example, we'll only set the `fillColor` and not actually paint anything using the `QPainter`.
+
+✅ Create a new `#[cxx_qt::bridge]` module
+
+It should include the following declarations/includes:
+* ✅ An import of `QColor` from cxx-qt-lib [(see the CXX-Qt documentation)](https://docs.rs/cxx-qt-lib/latest/cxx_qt_lib/struct.QColor.html)
+* ✅ An include to the `QQuickPaintedItem` header
+* ✅ An include for the `QPainter` header, as well as a declaration of the `QPainter` type.
+
+```rust
+#[cxx_qt::bridge(cxx_file_stem = "image_painter")]
+pub mod qobject {
+    unsafe extern "C++" {
+        include!("cxx-qt-lib/qcolor.h");
```

Comment:
maybe add to your questions below, `? why do we need to include the qcolor.h header` ? (which is a good question and i would be nice if we could somehow remove it but not sure we can bundle it into the type easily...)

Reply:
(assume we have checked recently that you do still need it :') )

Reply:
Hm, that's a good question.
Wonder if it's a bit advanced/hard to figure out from the documentation.

I'll put it in and mark it as (advanced).